### PR TITLE
Tsungyu Yu - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by ytyubox**
- Xcode 12.2 for the UI Design Patterns Challenge.
- 4th language in (Chinese, Traditional, zh-Hant).

This is my pull request for UI Design Patterns Challenge.
Please have a look, thanks. 😁

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+ *		
+ *		Using Swift 5.0
+ *		
+ *		Running on macOS 10.15
+ */
+
+
+import Foundation
+struct FeedErrorViewModel {
+	var error:Error?
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedErrorViewModel.swift:13]: The error could be an optional string `var error: String?`

This way, the UI doesn't decide which string to show. That's the presenter's responsibility.

File: MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift [hunk 2]
```diff
@@ -26,4 +27,12 @@ final class FeedViewAdapter: FeedView {
 			return view
 		})
 	}
+	
+	func display(_ viewModel: FeedErrorViewModel) {
+		let message: String? = viewModel.error.map{
+			_ in
+			Localized.Feed.loadError
+		}
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift:35]: It'd be best to perform this mapping in the Presenter. See my other comment.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 3]
```diff
@@ -20,12 +20,17 @@ final class FeedViewModel {
 	
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onFeedErroStateChange: Observer<Error?>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: The error could be an optional string `Observer<String?>`

This way, the UI doesn't decide which string to show. That's the ViewModel's responsibility.

File: MVVM Module/Source/Feed UI/Composers/FeedUIComposer.swift [hunk 4]
```diff
@@ -17,7 +17,20 @@ public final class FeedUIComposer {
...
+				: feedController?.errorView.show(message: Localized.Feed.loadError)
+		}
+		feedViewModel.onLoadingStateChange = {
+			[weak feedController] isLoading in
+			if isLoading {
+				feedController?.refreshControl?.beginRefreshing()
+			} else {
+				feedController?.refreshControl?.endRefreshing()
+			}
+		}
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Composers/FeedUIComposer.swift:33]: It'd be best to keep this binding logic in the view controller. The Composers should ideally only do composition, with no extra logic.

**Review by caiozullo — COMMENTED** [review]
Hi @ytyubox,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left some review comments. Feel free to push changes.

Great job!

**Comment by ytyubox** [comment]
@caiozullo Thanks for the feedback, I make a few changes based on your suggestion. Appreciated.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 5]
```diff
@@ -16,6 +17,22 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
...
+			}
+		}
+		viewModel?.onLoadingStateChange = {
+			[weak self] isLoading in
+			if isLoading {
+				self?.refreshControl?.beginRefreshing()
+			} else {
+				self?.refreshControl?.endRefreshing()
+			}
+		}
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:35]: This logic can be moved to the `bind` method that's called every time the view model is set. That's where it's safe to bind the views with the view model.

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 6]
```diff
@@ -11,6 +11,7 @@ protocol FeedLoadingView {
 
 protocol FeedView {
 	func display(_ viewModel: FeedViewModel)
+	func display(_ viewModel: FeedErrorViewModel)
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:14]: You can move this method to another protocol (e.g., FeedErrorView) and the view controller can implement it directly. This way, the Adapter wouldn't have to implement it to forward the message to the view controller (one less level of indirection!).

Comment (ytyubox) [reply MVP Module/Source/Feed Presentation/FeedPresenter.swift]: Sure, fix by ef88d30

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift (hunk 5 above)
Comment (ytyubox) [reply MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift]: Sure, ef88d30
then I refactor the closure injection into domain language as private function

File: MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift [hunk 7]
```diff
@@ -26,4 +27,8 @@ final class FeedViewAdapter: FeedView {
 			return view
 		})
 	}
+	
+	func display(_ viewModel: FeedErrorViewModel) {
+		controller?.display(viewModel)
+	}
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift:33]: This method in the adapter is not used anymore. You can remove it.

Comment (ytyubox) [inline MVP Module/Source/Feed UI/Composers/FeedViewAdapter.swift:33]: Sure, my bad
```suggestion
```

**Comment by caiozullo** [comment]
Nice one!

**Pull request closed without merging**